module-system/sov-state/src/vec.rs

```diff
@@ -0,0 +1,218 @@
+use std::iter::FusedIterator;
+use std::marker::PhantomData;
+
+use thiserror::Error;
+
+use crate::codec::{BorshCodec, PairOfCodecs, StateValueCodec};
+use crate::{Prefix, StateMap, StateValue, Storage, WorkingSet};
+
+type InternalCodec<C> = PairOfCodecs<BorshCodec, C>;
+
+#[derive(Debug, Clone)]
+pub struct StateVec<V, C = BorshCodec>
+where
+    C: StateValueCodec<V>,
+{
+    _phantom: PhantomData<V>,
+    prefix: Prefix,
+    len_value: StateValue<usize>,
+    elems: StateMap<usize, V, InternalCodec<C>>,
+}
+
+/// Error type for `StateVec` get method.
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Index out of bounds for index: {0}")]
+    IndexOutOfBounds(usize),
+    #[error("Value not found for prefix: {0} and index: {1}")]
+    MissingValue(Prefix, usize),
+}
+
+impl<V> StateVec<V>
+where
+    BorshCodec: StateValueCodec<V>,
+{
+    /// Crates a new [`StateVec`] with the given prefix and the default
+    /// [`StateValueCodec`] (i.e. [`BorshCodec`]).
+    pub fn new(prefix: Prefix) -> Self {
+        Self::with_codec(prefix, BorshCodec)
+    }
+}
+
+impl<V, C> StateVec<V, C>
+where
+    C: StateValueCodec<V>,
+{
+    /// Creates a new [`StateVec`] with the given prefix and codec.
+    pub fn with_codec(prefix: Prefix, codec: C) -> Self {
```

Comment:
It's a little surprising to me that we pass a codec *instance* here. Is there a reason for that?

Reply:
I'm not sure what the alternative would be. You mean a codec instance as opposed to only relying on the codec type `C` and storing `PhantomData<C>` internally? Codecs require `self` to encode and decode values, as we assume they may depend on some internal state or settings e.g. endianess, compression. It just so happens that most don't.